base/random.jl

```diff
@@ -1156,20 +1156,57 @@ function randexp_unlikely(rng, idx, x)
     end
 end
 
-function randexp!(rng::AbstractRNG, A::Array{Float64})
-    for i in eachindex(A)
-        @inbounds A[i] = randexp(rng)
-    end
-    A
-end
+"""
+    randn!([rng], A::Array{T,N})
 
-randexp!(A::Array{Float64}) = randexp!(GLOBAL_RNG, A)
-randexp(dims::Dims) = randexp!(Array(Float64, dims))
-randexp(dims::Int...) = randexp!(Array(Float64, dims))
-randexp(rng::AbstractRNG, dims::Dims) = randexp!(rng, Array(Float64, dims))
-randexp(rng::AbstractRNG, dims::Int...) = randexp!(rng, Array(Float64, dims))
+Fill the array `A` with normally-distributed (mean 0, standard deviation 1) random numbers.
+The type `T` can be `Float16`, `Float32`, or `Float64`.
+Also see the `rand` function.
+"""
+function randn! end
+
+"""
+    randexp!([rng], A::Array{T,N})
+
+Fill the array `A` with random numbers following the exponential distribution (with scale 1).
+The type `T` can be `Float16`, `Float32`, or `Float64`.
+"""
+function randexp! end
+
+let Floats = Union{Float16,Float32,Float64}
+    for randfun in [:randn, :randexp]
+        randfun! = Symbol(randfun, :!)
+        @eval begin
+
+            # scalars
 
+            $randfun{F<:$Floats}(rng::AbstractRNG, ::Type{F}) = convert(F, $randfun(rng))
+            $randfun{F<:$Floats}(::Type{F}) = $randfun(GLOBAL_RNG, F)
 
+            # filling arrays
+
+            function $randfun!{F<:$Floats}(rng::AbstractRNG, A::AbstractArray{F})
```

Comment:
I would use `T<:Number` here, to allow the possibility of e.g. `randn(::Complex)` being added in the future.